%\VignetteIndexEntry{motifStack Vignette}
%\VignetteDepends{motifStack}
%\VignetteKeywords{sequence logo}
%\VignettePackage{motifStack}
\documentclass[12pt]{article}
\usepackage{hyperref}
\usepackage{url}
\usepackage[numbers]{natbib}
\usepackage{graphicx}
\bibliographystyle{plainnat}

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\author{Jianhong Ou\footnote{jianhong.ou@umassmed.edu}, Lihua Julie Zhu\footnote{Julie.Zhu@umassmed.edu}}
\begin{document}
\title{motifStack guide}

\maketitle

\tableofcontents
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}
A sequence logo has been widely used as a graphical representation of an alignment of multiple amino acid or nucleic acid sequences.
There is a package seqlogo\cite{Oliver} implemented in R to draw DNA sequence logos. However, it does not support amino acid sequence logos.

We have developed motifStack package for drawing sequence logos for protein, DNA and RNA sequences. 
motifStack provides the flexibility for users to select the font type and symbol colors. 
Comparing to seqlogo, motifStack has the capability for graphical representation of multiple motifs.

\section{Prepare environment}
You will need ghostscript: the full path to the executable can be set by the environment variable R\_GSCMD. 
If this is unset, a GhostScript executable will be searched by name on your path. For example, on a Unix, linux or Mac "gs" is used for searching, 
and on Windows the setting of the environment variable GSC is used, otherwise commands "gswi64c.exe" then "gswin32c.exe" are tried.

Example on Windows: assume that the gswin32c.exe is installed at C:\textbackslash Program Files\textbackslash gs\textbackslash gs9.06\textbackslash bin, then open R and try:
\begin{scriptsize}
Sys.setenv(R\_GSCMD="\textbackslash"C:\textbackslash\textbackslash Program Files\textbackslash\textbackslash gs\textbackslash\textbackslash gs9.06\textbackslash\textbackslash bin\textbackslash\textbackslash gswin32c.exe\textbackslash"")
\end{scriptsize}

 
\section{Examples of using motifStack}
\subsection{plot a DNA sequence logo with different fonts and colors}
Users can select different fonts and colors to draw the sequence logo.
\begin{scriptsize}
<<DNAseqLogo,include=FALSE>>=
library(motifStack)
pcm <- read.table(file.path(find.package("motifStack"), "extdata", "bin_SOLEXA.pcm"))
pcm <- pcm[,3:ncol(pcm)]
rownames(pcm) <- c("A","C","G","T")
motif <- pcm2pfm(pcm)
motif <- new("pfm", mat=motif, name="bin_SOLEXA")
plot(motif)
#try a different font
plot(motif, font="mono,Courier")
#try a different font and a different color group
motif@color <- colorset(colorScheme='basepairing')
plot(motif,font="Times")
@
\end{scriptsize}
\begin{figure}[htb]
\centering
<<fig1,fig=TRUE,echo=FALSE>>=
opar<-par(mfrow=c(3,1))
motif@color<-colorset(colorScheme='auto')
motif@name="bin_SOLEXA, font='Helvetica', color='auto'"
plot(motif)
motif@name="bin_SOLEXA, font='mono,Courier', color='auto'"
plot(motif, font="mono,Courier")
motif@color <- colorset(colorScheme='basepairing')
motif@name="bin_SOLEXA, font='mono,Courier', color='basepairing'"
plot(motif,font="Times")
par<-opar
@
\caption{DNA sequence logo}
\label{fig:font}
\end{figure}

\subsection{plot an amino acid sequence logo}
Given that motifStack allows to use any letters as symbols, 
it can also be used to draw amino acid sequence logos.
\begin{scriptsize}
<<AAseqLogo,include=FALSE>>=
library(motifStack)
protein<-read.table(file.path(find.package("motifStack"),"extdata","cap.txt"))
protein<-t(protein[,1:20])
motif<-pcm2pfm(protein)
motif<-new("pfm", mat=motif, name="CAP", 
            color=colorset(alphabet="AA",colorScheme="chemistry"))
@
> plot(motif)
\end{scriptsize}
\begin{figure}[htb]
\centering
<<fig2,fig=TRUE,echo=FALSE>>=
plot(motif)
@
\caption{Amino acid sequence logo}
\label{fig:protein}
\end{figure}

\subsection{plot sequence logo stack}
motifStack is designed to show multiple motifs in same canvas. To show the sequence logo stack,
motifs need to be aligned first for example by using MotIV\cite{Eloi2010}::motifMatch, 
which implemented STAMP\cite{Mahony2007}. After alignment, users can use plotMotifLogoStack
function to draw sequence logos stack or use plotMotifLogoStackWithTree function to show the distance tree
and sequence logos stack in the same canvas.

\begin{scriptsize}
<<seqLogoStack,include=FALSE>>=
library(motifStack)
library("MotIV")
#####Database and Scores#####
jaspar <- MotIV::readPWMfile(file.path(find.package("MotIV"), "extdata", "jaspar2010.txt"))
jaspar.scores <- MotIV::readDBScores(file.path(find.package("MotIV"), "extdata", "jaspar2010_PCC_SWU.scores"))
#####Input#####
pcms<-readPCM(file.path(find.package("motifStack"), "extdata"),"pcm$")
pcms<-lapply(pcms,function(.ele){.ele<-.ele[,3:ncol(.ele)];rownames(.ele)<-c("A","C","G","T");.ele})
motifs<-lapply(pcms,pcm2pfm)

#####Analysis#####
foxa1.analysis.jaspar<-MotIV::motifMatch(inputPWM=motifs,
                                         align="SWU",cc="PCC",
                                         database=jaspar, DBscores=jaspar.scores,top=5)
#####Clustering#####
d <- MotIV::motifDistances(getPWM(foxa1.analysis.jaspar))
hc <- MotIV::motifHclust(d)

##reorder the motifs for plotMotifLogoStack
motifs<-motifs[hc$order]
motifs<-lapply(names(motifs), function(.ele, motifs){new("pfm",mat=motifs[[.ele]], name=.ele)},motifs)
##do alignment
motifs<-DNAmotifAlignment(motifs)
##plot stacks
@
> plotMotifLogoStack(motifs, ncex=1.0)
> plotMotifLogoStackWithTree(motifs, hc=hc)
\end{scriptsize}
\begin{figure}[htb]
\centering
<<fig3,fig=TRUE,echo=FALSE>>=
plotMotifLogoStack(motifs, ncex=1.0)
@
\caption{sequence logo stack}
\label{fig:logostack}
\end{figure}
\begin{figure}[htb]
\centering
<<fig4,fig=TRUE,echo=FALSE>>=
plotMotifLogoStackWithTree(motifs, hc=hc)
@
\caption{sequence logo stack with hierarchical cluster tree}
\label{fig:treestack}
\end{figure}

\subsection{plot sequence logo stack in radial style}
When the number of motifs is too much to be shown in a vertical stack, motifStack can draw them in a radial style.

\begin{scriptsize}
<<radialLogoStack,include=FALSE>>=
library("MotifDb")
library("MotIV")
library("motifStack")
matrix.fly <- query(MotifDb, "Dmelanogaster")
motifs <- as.list(matrix.fly)
motifs <- motifs[grepl("Dmelanogaster\\-FlyFactorSurvey\\-", names(motifs))]
names(motifs) <- gsub("Dmelanogaster_FlyFactorSurvey_", "", 
                      gsub("_FBgn\\d+$", "", 
                            gsub("[^a-zA-Z0-9]","_", 
                                 gsub("(_\\d+)+$", "", names(motifs)))))
motifs <- motifs[unique(names(motifs))]
pfms <- sample(motifs, 50)
jaspar <- MotIV::readPWMfile(file.path(find.package("MotIV"), "extdata", "jaspar2010.txt"))
jaspar.scores <- MotIV::readDBScores(file.path(find.package("MotIV"), "extdata", "jaspar2010_PCC_SWU.scores"))
ffs.analysis.jaspar<-MotIV::motifMatch(inputPWM=pfms,align="SWU",
                                       cc="PCC",database=jaspar, 
                                       DBscores=jaspar.scores,top=5)
d <- MotIV::motifDistances(getPWM(ffs.analysis.jaspar))
hc <- MotIV::motifHclust(d)
phylog <- hclust2phylog(hc)
leaves <- names(phylog$leaves)
pfms <- pfms[leaves]
pfms <- lapply(names(pfms), function(.ele, pfms){new("pfm",mat=pfms[[.ele]], name=.ele)},pfms)
pfms <- DNAmotifAlignment(pfms, minimalConsensus=3)
library(RColorBrewer)
color <- brewer.pal(12, "Set3")
@
> plotMotifStackWithRadialPhylog(phylog, pfms, circle=0.9, cleaves = 0.5, clabel.leaves = 0.7, 
                                 col.bg=rep(color, each=5), col.leaves=rep(color, each=5))
\end{scriptsize}
\begin{figure}[htb]
\centering
<<fig5,fig=TRUE,echo=FALSE>>=
plotMotifStackWithRadialPhylog(phylog, pfms, circle=0.3, cleaves = 0.2, clabel.leaves = 0.6, 
                                 col.bg=rep(color, each=5), col.leaves=rep(color, each=5))
@
\caption{sequence logo stack in radial style}
\label{fig:treestack}
\end{figure}

\section{References}
\begin{thebibliography}{99}
\bibitem[Oliver Bembom ()]{Oliver} seqLogo: Sequence logos for DNA sequence alignments. R package version 1.22.0. 
\bibitem[Eloi et al. (2010)]{Eloi2010} MotIV: Motif Identification and Validation. Eloi Mercier and Raphael Gottardo (2010). R package version 1.10.0.
\bibitem[Mahony et al. (2007)]{Mahony2007} STAMP: a web tool for exploring DNA-binding motif similarities. Mahony S, Benos PV, Nucleic Acids Res. 2007, 35(Web Server issue): W253-W258.
\end{thebibliography}

\section{Session Info}
<<>>=
sessionInfo()
@
\end{document}
